client/azure.go

```diff
@@ -123,83 +150,81 @@ func ResponseWasNotFound(resp autorest.Response) bool {
-		return nil, microerror.Maskf(err, "creating service principal token")
-	}
-
-	client := dns.NewZonesClient(config.SubscriptionID)
-	client.Authorizer = autorest.NewBearerAuthorizer(spt)
+func newDNSZonesClient(config *azureClientConfig) (*dns.ZonesClient, error) {
+	client := dns.NewZonesClientWithBaseURI(
+		config.resourceManagerEndpoint,
+		config.subscriptionID)
+	client.Authorizer = autorest.NewBearerAuthorizer(config.servicePrincipalToken)
 
 	return &client, nil
 }
 
-func newInterfacesClient(config AzureConfig) (*network.InterfacesClient, error) {
-	spt, err := newServicePrincipalToken(config, azure.PublicCloud.ResourceManagerEndpoint)
-	if err != nil {
-		return nil, microerror.Maskf(err, "creating service principal token")
-	}
-
-	client := network.NewInterfacesClient(config.SubscriptionID)
-	client.Authorizer = autorest.NewBearerAuthorizer(spt)
+func newInterfacesClient(config *azureClientConfig) (*network.InterfacesClient, error) {
+	client := network.NewInterfacesClientWithBaseURI(
+		config.resourceManagerEndpoint,
+		config.subscriptionID)
+	client.Authorizer = autorest.NewBearerAuthorizer(config.servicePrincipalToken)
 
 	return &client, nil
 }
 
-func newVnetPeeringClient(config AzureConfig) (*network.VirtualNetworkPeeringsClient, error) {
-	spt, err := newServicePrincipalToken(config, azure.PublicCloud.ResourceManagerEndpoint)
-	if err != nil {
-		return nil, microerror.Maskf(err, "creating service principal token")
-	}
-
-	client := network.NewVirtualNetworkPeeringsClient(config.SubscriptionID)
-	client.Authorizer = autorest.NewBearerAuthorizer(spt)
+func newVnetPeeringClient(config *azureClientConfig) (*network.VirtualNetworkPeeringsClient, error) {
+	client := network.NewVirtualNetworkPeeringsClientWithBaseURI(
+		config.resourceManagerEndpoint,
+		config.subscriptionID)
+	client.Authorizer = autorest.NewBearerAuthorizer(config.servicePrincipalToken)
 
 	return &client, nil
 }
 
-func newServicePrincipalToken(config AzureConfig, scope string) (*adal.ServicePrincipalToken, error) {
-	oauthConfig, err := adal.NewOAuthConfig(azure.PublicCloud.ActiveDirectoryEndpoint, config.TenantID)
+func newServicePrincipalToken(config AzureConfig, env *azure.Environment) (*adal.ServicePrincipalToken, error) {
+	oauthConfig, err := adal.NewOAuthConfig(env.ActiveDirectoryEndpoint, config.TenantID)
 	if err != nil {
 		return nil, microerror.Maskf(err, "creating OAuth config")
 	}
 
-	return adal.NewServicePrincipalToken(*oauthConfig, config.ClientID, config.ClientSecret, scope)
+	return adal.NewServicePrincipalToken(
+		*oauthConfig,
+		config.ClientID,
+		config.ClientSecret,
+		env.ServiceManagementEndpoint)
+}
+
+// parseAzureEnvironment returns azure environment by name.
+func parseAzureEnvironment(cloudName string) (*azure.Environment, error) {
+	var env azure.Environment
+	var err error
+	if cloudName == "" {
+		env = azure.PublicCloud
+	} else {
+		env, err = azure.EnvironmentFromName(cloudName)
+	}
+	return &env, err
```

Comment:
There should be error masking and I would question the need of returning a pointer here.

Reply:
done. Returning value and masked error in this function and wrpa error above, where function called.